core/client/src/main/java/alluxio/client/block/stream/NettyPacketReader.java

```diff
@@ -0,0 +1,400 @@
+  }
+
+  @Override
+  public long pos() {
+    return mPosToRead;
+  }
+
+  @Override
+  public ByteBuf readPacket() throws IOException {
+    Preconditions.checkState(!mClosed, "PacketReader is closed while reading packets.");
+    ByteBuf buf = null;
+    mLock.lock();
+    try {
+      while (true) {
+        if (mDone) {
+          return null;
+        }
+        if (mPacketReaderException != null) {
+          throw new IOException(mPacketReaderException);
+        }
+        buf = mPackets.poll();
+        if (!tooManyPacketsPending()) {
+          resume();
+        }
+        if (buf == null) {
+          try {
+            if (!mNotEmptyOrFail.await(READ_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+              throw new IOException(
+                  String.format("Timeout while reading packet from block %d @ %s.", mId, mAddress));
+            }
+          } catch (InterruptedException e) {
+            throw Throwables.propagate(e);
+          }
+        } else {
+          if (buf.readableBytes() == 0) {
+            buf.release();
+            mDone = true;
+            return null;
+          }
+          mPosToRead += buf.readableBytes();
+          Preconditions.checkState(mPosToRead - mStart <= mBytesToRead);
+          return buf;
+        }
+      }
+    } catch (Throwable e) {
+      if (buf != null) {
+        buf.release();
+      }
+      throw e;
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  @Override
+  public void close() {
+    try {
+      if (mDone) {
+        return;
+      }
+      if (!mChannel.isOpen()) {
+        return;
+      }
+      try {
+        if (!CANCEL_ENABLED) {
+          mChannel.close().sync();
+          return;
+        }
+        if (remaining() > 0) {
+          Protocol.ReadRequest cancelRequest =
+              Protocol.ReadRequest.newBuilder().setId(mId).setCancel(true).setType(mRequestType)
+                  .build();
+          mChannel.writeAndFlush(new RPCProtoMessage(cancelRequest))
+              .addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
+        }
+      } catch (InterruptedException e) {
+        mChannel.close();
+        throw Throwables.propagate(e);
+      }
+
+      while (true) {
+        try {
+          ByteBuf buf = readPacket();
+          // A null packet indicates the end of the stream.
+          if (buf == null) {
+            return;
+          }
+          buf.release();
+        } catch (IOException e) {
+          LOG.warn("Failed to close the NettyBlockReader (block: {}, address: {}).",
+              mId, mAddress, e);
+          try {
+            mChannel.close().sync();
+          } catch (InterruptedException ee) {
+            throw Throwables.propagate(ee);
+          }
+          return;
+        }
+      }
+    } finally {
+      if (mChannel.isOpen()) {
+        Preconditions.checkState(mChannel.pipeline().last() instanceof Handler);
+        mChannel.pipeline().removeLast();
+        resume();
```

Comment:
`resume` is a bit confusing, is it called because we are trying to restore the state of `autoread` for the channel? Perhaps we should have something like `removeHandler` to make it clear that the channel is temporarily being modified for our use and restored to the previous state afterward